[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77288

Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

Differential Revision: [D36347737](https://our.internmc.facebook.com/intern/diff/D36347737)

VitalyFedyunin committed:
[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"




[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -265,3 +278,43 @@ def __iter__(self) -> Iterator[T]:
         for i in self.items:
             yield i
+
+class TravelingDataPipe:
```
Could we rename it to something like `_DataPipeSerializationWrapper`? With `DataPipe` at the end of this class, we might give users a wrong hint. And, make it private (underscore) might be even better.

NivekT replied:
I agree with the renaming, we might want to rename `TravelingIterDataPipe` and `TravelingMapDataPipe` too, especially if these names show up in the DataPipe graph and users want to know what they are

ejguan reviewed (approved):
LGTM with a couple comments below!

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"


Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"


Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"


Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

[ghstack-poisoned]

ejguan commented on test/test_datapipe.py:
```diff
@@ -2082,6 +2082,23 @@ def test_traverse_mixdatapipe(self):
+    return x[0]
+
+class TestSerialization(TestCase):
+    @skipIfNoDill
```
Lol. I completely mess up the `spawn` and `fork`. Sorry abotu that

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"


Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

Differential Revision: [D36347737](https://our.internmc.facebook.com/intern/diff/D36347737)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"


Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

Differential Revision: [D36347737](https://our.internmc.facebook.com/intern/diff/D36347737)

[ghstack-poisoned]

NivekT commented:
I think users can also use the same pattern here if they ever want to have custom serialization logic? I am anticipating that they can follow this pattern, implement a DataPipe with custom `__getstate__` and `__setstate__`, put that at the end of the pipeline, and the rest should work.

VitalyFedyunin committed:
Update on "[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing"


Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

Differential Revision: [D36347737](https://our.internmc.facebook.com/intern/diff/D36347737)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36347737).

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36347737).

VitalyFedyunin commented on torch/utils/data/communication/eventloop.py:
```diff
@@ -44,8 +55,15 @@ def SpawnThreadForDataPipeline(datapipe):
+            try:
+                new_datapipe = dill.loads(dill.dumps(datapipe))
+            except Exception as de:
+                Exception('Unable to dill DataPipe to make thread local copy', de)
```
Good catch, I will follow-up with quick patch.
